### PR TITLE
Optimize hasContainers addition by ensuring correct size

### PR DESCRIPTION
**Pull request by porunov**
Related to: https://github.com/JanusGraph/janusgraph/pull/2660#discussion_r650742639

Signed-off-by: Oleksandr Porunov <alexandr.porunov@gmail.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphPropertiesStep.java [hunk 1]
```diff
@@ -189,14 +188,24 @@ public void reset() {
...
     @Override
-    public void addAll(Iterable<HasContainer> has) {
-        Iterables.addAll(hasContainers, has);
+    public int getHasContainersAmount() {
+        return hasContainers.size();
+    }
+
+    @Override
+    public void ensureHasContainersCapacity(int size) {
+        hasContainers.ensureCapacity(size);
```
Comment (rngcntr) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphPropertiesStep.java:203]: Nitpick: This method could also be called `ensureAdditionalHasContainersCapacity(int additionalSize)` and add `hasContainers.size()` on its own. By that, we could avoid the extra method `getHasContainersAmount()`.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/HasStepFolder.java [hunk 2]
```diff
@@ -303,15 +314,20 @@ static Step foldInHasFilter(final Traversal.Admin<?, ?> traversal, final Travers
         return currentStep;
     }
 
+    static void splitAndP(final List<HasContainer> hasContainers, final HasContainer hasContainer) {
+        if (hasContainer.getPredicate() instanceof AndP) {
+            for (final P<?> predicate : ((AndP<?>) hasContainer.getPredicate()).getPredicates()) {
+                hasContainers.add(new HasContainer(hasContainer.getKey(), predicate));
```
[the remark this answers is not in the document]
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/HasStepFolder.java:320]: Checking this closer. This loop cannot be triggered at all. 
Before we entering this loop we actually convert all `HasContainers` via `JanusGraphPredicateUtils.convert`. In that method we will convert any has container which has a predicate `AndP` and `OrP` into `ConnectiveJanusGraphP` which doesn't extend `AndP`.
So, it looks like the only way to enter this `if` condition is via tests (call `addHasContainer` with `HasContainer` which is instance of `AndP`) but generally in runtime we won't ever trigger this `if` condition.
Now I wonder if we need this loop at all. Probably we should simply remove this method.

[the remark this answers is not in the document]
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/HasStepFolder.java:320]: Or maybe, if anyone manually insert `HasContainer` into some `JanusGraphStep` and don't convert that `HasContainer` via `JanusGraphPredicateUtils.convert` than this code could be triggered. That said, I'm not sure if anyone does it. Probably it's still better to leave `splitAndP` but just cover such an unusual usage I guess.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/HasStepFolder.java [hunk 3]
```diff
@@ -303,15 +314,20 @@ static Step foldInHasFilter(final Traversal.Admin<?, ?> traversal, final Travers
...
+            }
+        } else {
+            hasContainers.add(hasContainer);
+        }
+    }
+
     static List<HasContainer> splitAndP(final List<HasContainer> hasContainers, final Iterable<HasContainer> has) {
```
[the remark this answers is not in the document]
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/HasStepFolder.java]: Removed this method

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphPropertiesStep.java (hunk 1 above)
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphPropertiesStep.java]: Changed to `ensureAdditionalHasContainersCapacity` and removed `getHasContainersAmount`

**Review by rngcntr — APPROVED** [review]

**Pull request merged**